File: SGUEES-SPA/src/app/pages/SelectionHiring/sc-competencias-conductuales/sc-competencias-conductuales.service.ts
// Qué hace: servicio de negocio del catálogo Competencias Conductuales.
// Cómo: valida los datos, ejecuta el CRUD a través del repositorio y arma la configuración de grilla y formulario.
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { IParam } from 'src/app/FxAPI/IParam';
import { IResult } from 'src/app/FxAPI/IResult';
import { NotifyType } from 'src/app/shared/models/NotifyType';
import { createEstadoColumnConfig, ESTADO_ACTIVO_INACTIVO_LABELS } from 'src/app/shared/utils/remote-grid-filter.util';
import { buildAuditGridColumns } from 'src/app/shared/mtto/mtto-grid.helpers';
import { ScCompetenciasConductuales } from './models/sc-competencias-conductuales';
import { ScCompetenciasConductualesRepository } from './sc-competencias-conductuales.repository';

const ESTADO_FIELD = 'ESTADO_COMPETENCIAS_CONDUCTUALES';

@Injectable({ providedIn: 'root' })
// Qué hace: servicio de competencias conductuales.
// Cómo: valida los datos y llama a ScCompetenciasConductualesRepository para ejecutar el CRUD.
export class ScCompetenciasConductualesService {
	constructor(private repo: ScCompetenciasConductualesRepository) {}

	// Qué hace: valida el formulario de competencia conductual antes de guardar.
	// Cómo: revisa CORR_TIPO_PUESTO, NOMBRE_COMPETENCIAS_CONDUCTUALES y DESCRIPCION, notificando con msg cuando falla.
	esValido(model: ScCompetenciasConductuales, msg: Function): boolean {
		if (!model.CORR_TIPO_PUESTO || model.CORR_TIPO_PUESTO <= 0) {
			msg('Debe seleccionar el tipo de puesto.', NotifyType.Warning);
			return false;
		}

		if (!model.NOMBRE_COMPETENCIAS_CONDUCTUALES || model.NOMBRE_COMPETENCIAS_CONDUCTUALES.trim() === '') {
			msg('Debe ingresar el nombre de la competencia conductual.', NotifyType.Warning);
			return false;
		}

		if (model.NOMBRE_COMPETENCIAS_CONDUCTUALES.trim().length > 150) {
			msg('El nombre de la competencia conductual no puede superar 150 caracteres.', NotifyType.Warning);
			return false;
		}

		if (!model.DESCRIPCION || model.DESCRIPCION.trim() === '') {
			msg('Debe ingresar la descripcion de la competencia conductual.', NotifyType.Warning);
			return false;
		}

		if (model.DESCRIPCION.trim().length > 500) {
			msg('La descripcion no puede superar 500 caracteres.', NotifyType.Warning);
			return false;
		}

		return true;
	}

	// Qué hace: obtiene el listado de competencias conductuales.
	// Cómo: llama a getAll del repositorio con el filtro construido por buildWhere.
	getAll(param: any): Observable<IResult> {
		return this.repo.getAll(this.buildWhere(param));
	}

	// Qué hace: obtiene una competencia conductual puntual.
	// Cómo: llama a get del repositorio filtrando por CORR_COMPETENCIAS_CONDUCTUALES.
	get(param: any): Observable<IResult> {
		return this.repo.get([{ Parameter: 'CORR_COMPETENCIAS_CONDUCTUALES', Value: param.CORR_COMPETENCIAS_CONDUCTUALES }]);
	}

	// Qué hace: crea una nueva competencia conductual.
	// Cómo: llama a create del repositorio con el modelo recibido.
	insert(model: any): Observable<IResult> {
		return this.repo.create(model);
	}

	// Qué hace: actualiza una competencia conductual existente.
	// Cómo: llama a update del repositorio con el modelo y su CORR_COMPETENCIAS_CONDUCTUALES.
	update(model: any): Observable<IResult> {
		return this.repo.update(model, [
			{ Parameter: 'CORR_COMPETENCIAS_CONDUCTUALES', Value: model.CORR_COMPETENCIAS_CONDUCTUALES },
		]);
	}

	// Qué hace: elimina una competencia conductual.
	// Cómo: llama a delete del repositorio filtrando por CORR_COMPETENCIAS_CONDUCTUALES.
	delete(model: any): Observable<IResult> {
		return this.repo.delete([{ Parameter: 'CORR_COMPETENCIAS_CONDUCTUALES', Value: model.CORR_COMPETENCIAS_CONDUCTUALES }]);
	}

	// Qué hace: cambia el estado activo/inactivo de una competencia conductual.
	// Cómo: llama a activarInactivar del repositorio filtrando por CORR_COMPETENCIAS_CONDUCTUALES.
	activarInactivar(model: any): Observable<IResult> {
		return this.repo.activarInactivar(model, [
			{ Parameter: 'CORR_COMPETENCIAS_CONDUCTUALES', Value: model.CORR_COMPETENCIAS_CONDUCTUALES },
		]);
	}

	// Qué hace: define columnas y formatos de la grilla de mantenimiento.
	// Cómo: arma el arreglo de columnas (correlativo, nombre, descripción, tipo puesto, estado y auditoría) usado por app-data-grid-mtto.
	getColumns(): any {
		return [
			{
				dataField: 'CORR_COMPETENCIAS_CONDUCTUALES',
				caption: 'Corr.',
				width: 90,
				dataType: 'number',
				filterOperations: ['=', '<', '>', '<=', '>='],
			},
			{ dataField: 'NOMBRE_COMPETENCIAS_CONDUCTUALES', caption: 'Nombre', width: 250 },
			{ dataField: 'DESCRIPCION', caption: 'Descripcion', width: 280 },
			{ dataField: 'NOMBRE_TIPO_PUESTO', caption: 'Tipo Puesto', width: 200 },
			createEstadoColumnConfig(ESTADO_FIELD, ESTADO_ACTIVO_INACTIVO_LABELS),
			...buildAuditGridColumns({ withDateTimeFilter: true }),
		];
	}

	// Qué hace: configura el contador de registros de la grilla.
	// Cómo: define el resumen totalItems que cuenta CORR_COMPETENCIAS_CONDUCTUALES.
	getSummary(): any {
		return {
			totalItems: [
				{
					column: 'CORR_COMPETENCIAS_CONDUCTUALES',
					summaryType: 'count',
					valueFormat: '#,##0',
					displayFormat: 'Cant: {0}',
				},
			],
		};
	}

	// Qué hace: define los campos y reglas del formulario de competencia conductual.
	// Cómo: arma el arreglo de items (correlativo, tipo puesto, nombre, estado y descripción) usado por dx-form.
	getItems(): any {
		return [
			{
				dataField: 'CORR_COMPETENCIAS_CONDUCTUALES',
				label: { text: 'Corr.' },
				colSpan: 1,
				editorOptions: { readOnly: true },
			},
			{
				dataField: 'CORR_TIPO_PUESTO',
				label: { text: 'Tipo Puesto' },
				colSpan: 2,
				editorOptions: { placeholder: 'Seleccione tipo de puesto...', showClearButton: false },
				template: 'CORR_TIPO_PUESTOLookup',
				validationRules: [
					{
						type: 'custom',
						message: 'Este campo es obligatorio',
						reevaluate: true,
						validationCallback: (e: { value: unknown }) => {
							const value = Number(e.value);
							return !Number.isNaN(value) && value > 0;
						},
					},
				],
			},
			{
				dataField: 'NOMBRE_COMPETENCIAS_CONDUCTUALES',
				label: { text: 'Nombre' },
				colSpan: 3,
				editorOptions: { placeholder: 'Nombre competencia conductual...', showClearButton: true, maxLength: 150 },
				validationRules: [{ type: 'required', message: 'Este campo es obligatorio' }],
			},
			{
				dataField: 'ESTADO_COMPETENCIAS_CONDUCTUALES',
				label: { text: 'Activo' },
				editorType: 'dxCheckBox',
				colSpan: 2,
			},
			{
				dataField: 'DESCRIPCION',
				label: { text: 'Descripcion' },
				colSpan: 8,
				editorType: 'dxTextArea',
				editorOptions: { placeholder: 'Descripcion...', showClearButton: true, maxLength: 500, height: 90 },
				validationRules: [{ type: 'required', message: 'Este campo es obligatorio' }],
			},
		];
	}

	// Qué hace: traduce los filtros del componente al formato esperado por la API.
	// Cómo: agrega a xWhere el parámetro CORR_COMPETENCIAS_CONDUCTUALES cuando viene informado en param.
	private buildWhere(param: any): IParam[] {
		const xWhere: IParam[] = [];

		if (param.CORR_COMPETENCIAS_CONDUCTUALES) {
			xWhere.push({ Parameter: 'CORR_COMPETENCIAS_CONDUCTUALES', Value: param.CORR_COMPETENCIAS_CONDUCTUALES });
		}

		return xWhere;
	}
}

File: SGUEES-API/sguees.api/Options/SelectionHiring/SC_COMPETENCIAS_CONDUCTUALES/SC_COMPETENCIAS_CONDUCTUALESService.cs
// Qué hace: lógica de negocio del catálogo competencias conductuales.
// Cómo: valida los datos y llama a ISC_COMPETENCIAS_CONDUCTUALESRepository para ejecutar el CRUD.
using System.Collections.Generic;
using System.Threading.Tasks;
using eFramework.Core;
using SGUEES.Models;
using SGUEES.Repositories;

namespace SGUEES.Services
{
  // Qué hace: servicio de competencias conductuales.
  // Cómo: valida los datos y llama al repositorio para persistir la información.
  public class SC_COMPETENCIAS_CONDUCTUALESService : ISC_COMPETENCIAS_CONDUCTUALESService
  {
    private readonly ISC_COMPETENCIAS_CONDUCTUALESRepository _repo;

    public SC_COMPETENCIAS_CONDUCTUALESService(ISC_COMPETENCIAS_CONDUCTUALESRepository repo)
    {
      _repo = repo;
    }

    // Qué hace: obtiene el listado de competencias conductuales.
    // Cómo: llama a GetAllAsync del repositorio con los parámetros armados por BuildParameters.
    public async Task<CResult> GetAllAsync(SC_COMPETENCIAS_CONDUCTUALESParam xWhere)
    {
      return await _repo.GetAllAsync(BuildParameters(xWhere));
    }

    // Qué hace: obtiene una competencia conductual puntual.
    // Cómo: llama a GetAsync del repositorio filtrando por CORR_EMPRESA y CORR_COMPETENCIAS_CONDUCTUALES.
    public async Task<CResult> GetAsync(SC_COMPETENCIAS_CONDUCTUALESParam xWhere)
    {
      var p = new List<CParameter>
      {
        new CParameter() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
        new CParameter() { ParameterName = "CORR_COMPETENCIAS_CONDUCTUALES", Value = xWhere.CORR_COMPETENCIAS_CONDUCTUALES, DbType = System.Data.DbType.Int32 },
      };

      return await _repo.GetAsync(p);
    }

    // Qué hace: crea una competencia conductual.
    // Cómo: valida la empresa de sesión y los datos con ValidateEmpresaSesion y Validate, normaliza con NormalizeData y llama a CreateAsync del repositorio.
    public async Task<CResult> CreateAsync(SC_COMPETENCIAS_CONDUCTUALESTable Data, string vLOGIN_SISTEMA, string vESTACION)
    {
      var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
      if (empresaError != null)
      {
        return empresaError;
      }

      var validation = Validate(Data);
      if (validation != null)
      {
        return validation;
      }

      NormalizeData(Data);
      return await _repo.CreateAsync(Data, vLOGIN_SISTEMA, vESTACION);
    }

    // Qué hace: actualiza una competencia conductual.
    // Cómo: valida la empresa de sesión, los datos y el CORR_COMPETENCIAS_CONDUCTUALES a actualizar, normaliza con NormalizeData y llama a UpdateAsync del repositorio.
    public async Task<CResult> UpdateAsync(SC_COMPETENCIAS_CONDUCTUALESTable Data, string vLOGIN_SISTEMA, string vESTACION)
    {
      var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
      if (empresaError != null)
      {
        return empresaError;
      }

      var validation = Validate(Data);
      if (validation != null)
      {
        return validation;
      }

      if (Data.CORR_COMPETENCIAS_CONDUCTUALES <= 0)
      {
        return ValidationError("No se pudo identificar la competencia conductual a actualizar.");
      }

      NormalizeData(Data);
      return await _repo.UpdateAsync(Data, vLOGIN_SISTEMA, vESTACION);
    }

    // Qué hace: elimina una competencia conductual.
    // Cómo: valida la empresa de sesión con ValidateEmpresaSesion y llama a DeleteAsync del repositorio.
    public async Task<CResult> DeleteAsync(SC_COMPETENCIAS_CONDUCTUALESTable Data, string vLOGIN_SISTEMA, string vESTACION)
    {
      var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
      if (empresaError != null)
      {
        return empresaError;
      }

      return await _repo.DeleteAsync(Data, vLOGIN_SISTEMA, vESTACION);
    }

    // Qué hace: cambia el estado activo/inactivo de una competencia conductual.
    // Cómo: valida la empresa de sesión y el CORR_COMPETENCIAS_CONDUCTUALES a actualizar, y llama a ActivarInactivarAsync del repositorio.
    public async Task<CResult> ActivarInactivarAsync(SC_COMPETENCIAS_CONDUCTUALESTable Data, string vLOGIN_SISTEMA, string vESTACION)
    {
      var empresaError = ValidateEmpresaSesion(Data.CORR_EMPRESA);
      if (empresaError != null)
      {
        return empresaError;
      }

      if (Data.CORR_COMPETENCIAS_CONDUCTUALES <= 0)
      {
        return ValidationError("No se pudo identificar la competencia conductual a actualizar.");
      }

      return await _repo.ActivarInactivarAsync(Data, vLOGIN_SISTEMA, vESTACION);
    }

    // Qué hace: arma los parámetros de filtro para el repositorio.
    // Cómo: construye la lista con CORR_EMPRESA a partir de xWhere.
    private static List<CParameter> BuildParameters(SC_COMPETENCIAS_CONDUCTUALESParam xWhere)
    {
      return new List<CParameter>
      {
        new CParameter() { ParameterName = "CORR_EMPRESA", Value = xWhere.CORR_EMPRESA, DbType = System.Data.DbType.Int32 },
      };
    }

    // Qué hace: normaliza los datos antes de guardar.
    // Cómo: recorta espacios de NOMBRE_COMPETENCIAS_CONDUCTUALES y DESCRIPCION, y aplica ESTADO_COMPETENCIAS_CONDUCTUALES activo cuando no viene informado.
    private static void NormalizeData(SC_COMPETENCIAS_CONDUCTUALESTable Data)
    {
      Data.NOMBRE_COMPETENCIAS_CONDUCTUALES = Data.NOMBRE_COMPETENCIAS_CONDUCTUALES?.Trim();
      Data.DESCRIPCION = Data.DESCRIPCION?.Trim();
      Data.ESTADO_COMPETENCIAS_CONDUCTUALES ??= true;
    }

    // Qué hace: valida los datos obligatorios de la competencia conductual.
    // Cómo: comprueba que Data no sea nulo, que CORR_TIPO_PUESTO sea válido, y que NOMBRE_COMPETENCIAS_CONDUCTUALES y DESCRIPCION no estén vacíos ni superen sus límites.
    private static CResult Validate(SC_COMPETENCIAS_CONDUCTUALESTable Data)
    {
      if (Data == null)
      {
        return ValidationError("No se recibieron datos de la competencia conductual.");
      }

      if (!Data.CORR_TIPO_PUESTO.HasValue || Data.CORR_TIPO_PUESTO.Value <= 0)
      {
        return ValidationError("Debe seleccionar el tipo de puesto.");
      }

      if (string.IsNullOrWhiteSpace(Data.NOMBRE_COMPETENCIAS_CONDUCTUALES))
      {
        return ValidationError("Debe ingresar el nombre de la competencia conductual.");
      }

      if (Data.NOMBRE_COMPETENCIAS_CONDUCTUALES.Trim().Length > 150)
      {
        return ValidationError("El nombre de la competencia conductual no puede superar 150 caracteres.");
      }

      if (string.IsNullOrWhiteSpace(Data.DESCRIPCION))
      {
        return ValidationError("Debe ingresar la descripcion de la competencia conductual.");
      }

      if (Data.DESCRIPCION.Trim().Length > 500)
      {
        return ValidationError("La descripcion no puede superar 500 caracteres.");
      }

      return null;
    }

    // Qué hace: obtiene las competencias conductuales activas para el descriptor de puesto.
    // Cómo: llama a GetCatalogoDescriptorAsync del repositorio filtrando por CORR_EMPRESA y arma el CResult con el listado.
    public async Task<CResult> GetCatalogoDescriptorAsync(SC_COMPETENCIAS_CONDUCTUALESParam xWhere)
    {
      var rows = await _repo.GetCatalogoDescriptorAsync(xWhere.CORR_EMPRESA);
      return new CResult
      {
        Data = rows,
        Result = true,
        CodeHelper = 0,
        ErrorCode = 0,
        ErrorMessage = "",
        ErrorSource = "",
        RowsAffected = rows.Count,
      };
    }

    // Qué hace: verifica que exista empresa en la sesión.
    // Cómo: si corrEmpresa es mayor a cero devuelve null; de lo contrario devuelve un CResult con el error de empresa no asignada.
    private static CResult ValidateEmpresaSesion(int corrEmpresa)
    {
      if (corrEmpresa > 0)
      {
        return null;
      }

      return new CResult
      {
        Data = null,
        Result = false,
        CodeHelper = 0,
        ErrorCode = 4100,
        ErrorMessage = "No se pudo guardar la competencia conductual porque su usuario no tiene una empresa asignada. Solicite que le configuren una empresa por defecto en el sistema.",
        ErrorSource = "[SC_COMPETENCIAS_CONDUCTUALESService]",
        RowsAffected = 0
      };
    }

    // Qué hace: construye un resultado de error de validación.
    // Cómo: arma un CResult con Result en false y el mensaje recibido.
    private static CResult ValidationError(string message)
    {
      return new CResult
      {
        Data = null,
        Result = false,
        CodeHelper = 0,
        ErrorCode = -1,
        ErrorMessage = message,
        ErrorSource = "[SC_COMPETENCIAS_CONDUCTUALESService]",
        RowsAffected = 0
      };
    }
  }
}
